Allow using env var to specify pidfile

benthorner opened the pull request:
### Summary

Previously it was only possible to specify the location of the pidfile
for the 'rails server' command with the '-P' flag. This adds support for
specifying the pidfile using a PIDFILE env var, which can still be
overridden by the '-P' flag and with the default pidfile path unchanged.

The motivation for this feature comes from using Docker to run multiple
instances of the same rails app. When developing a rails app with
Docker, it's common to bind-mount the rails root directory in the
running container, so that changes to files are shared between the
container and the host. However, this doesn't work so well with the
pidfile and it's necessary to (remember to) add a '-P' flag to the
'rails server' command line; being able to specify this flag using an
env var would make developing with Rails+Docker a bit simpler.

<!-- Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together. -->

### Other Information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

Commit:
Allow using env var to specify pidfile

Previously it was only possible to specify the location of the pidfile
for the 'rails server' command with the '-P' flag. This adds support for
specifying the pidfile using a PIDFILE env var, which can still be
overridden by the '-P' flag and with the default pidfile path unchanged.

The motivation for this feature comes from using Docker to run multiple
instances of the same rails app. When developing a rails app with
Docker, it's common to bind-mount the rails root directory in the
running container, so that changes to files are shared between the
container and the host. However, this doesn't work so well with the
pidfile and it's necessary to (remember to) add a '-P' flag to the
'rails server' command line; being able to specify this flag using an
env var would make developing with Rails+Docker a bit simpler.

y-yagi commented on railties/CHANGELOG.md:
```diff
@@ -1,3 +1,5 @@
+# (Unreleased)
 
+   - Support using environment variable to set pidfile
```
Please use `*` and add your name. Please refer to other CHANGELOG files.
https://github.com/rails/rails/blob/master/activesupport/CHANGELOG.md
https://github.com/rails/rails/blob/master/activerecord/CHANGELOG.md

y-yagi commented on railties/lib/rails/commands/server/server_command.rb:
```diff
@@ -114,7 +114,7 @@ class ServerCommand < Base # :nodoc:
         desc: "Runs server as a Daemon."
       class_option :using, aliases: "-u", type: :string,
         desc: "Specifies the Rack server used to run the application (thin/puma/webrick).", banner: :name
-      class_option :pid, aliases: "-P", type: :string, default: DEFAULT_PID_PATH,
```
This change affected to an output of help. Please add a default value to desc like `port` and `binding`.

y-yagi commented:
You should need to fix `prepare_restart`  also.
https://github.com/rails/rails/blob/c8a84603e2fa97822407bf4a1c8ac0308a899163/railties/lib/rails/commands/server/server_command.rb#L263-L265

benthorner commented:
@y-yagi thanks for your comments - first time working on Rails and I missed a few things :-). I've addressed them all - couldn't find a particular test for the `prepare_restart` method, but I changed it to use the `pid` method now.

y-yagi commented:
I missed to point out `user_supplied_options`. Can you fix to `user_supplied_options` when`PIDFILE` env is specified?
  
https://github.com/rails/rails/blob/aeba121a83965d242ed6d7fd46e9c166079a3230/railties/lib/rails/commands/server/server_command.rb#L208-L210

benthorner commented:
@y-yagi sure, done :-).

y-yagi commented:
@benthorner Thanks!!
